exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,257 @@
+  double perimeter() override { return 4 * width; }
+
+  // Calculates and returns the area
+  double area() override { return width * width; }
+};
+
+/**
+ *  Rectangle class
+ **/
+class Rectangle : public Shape {
+
+// Own side lengths
+private:
+  double width;
+  double height;
+
+// Method implementations
+public:
+  // Constructor
+  Rectangle(double widthLen, double heightLen) : Shape("Rectangle", 4) {
+    width = widthLen;
+    height = heightLen;
+  }
+
+  // Calculates and returns the perimeter
+  double perimeter() override { return (2 * width + 2 * height); }
+
+  // Calculates and returns the area
+  double area() override { return width * height; }
+};
+
+/**
+ *  Circle class
+ **/
+class Circle : public Shape {
+
+// Own radius
+private:
+  double radius;
+
+// Method implementations
+public:
+  // Constructor
+  Circle(double radiusLen) : Shape("Circle", 0) { radius = radiusLen; }
+
+  // Calculates and returns the perimeter
+  double perimeter() override { return (2 * M_PI * radius); }
+
+  // Calculates and returns the area
+  double area() override { return (M_PI * radius * radius); }
+};
+
+/**
+ *  Triangle class
+ **/
+class Triangle : public Shape {
+
+// Own height and base
+private:
+  double height;
+  double base;
+
+// Method implementations
+public:
+  // Constructor
+  Triangle(double heightLen, double baseLen) : Shape("Triangle", 3) {
+    height = heightLen;
+    base = baseLen;
+  }
+
+  // Calculates and returns the perimeter
+  double perimeter() {
+    return (base + 2 * sqrt(height * height + (base * base) / 4));
+  }
+
+  // Calculates and returns the area
+  double area() { return (0.5 * base * height); }
+};
+
+/**
+ *  ShapeSorter class
+ **/
+class ShapeSorter {
+
+  /**
+   * Comparators for sort()
+   *
+   * http://www.cplusplus.com/reference/algorithm/sort/
+   **/
+  static bool compareArea(Shape *shape1, Shape *shape2) {
+    return shape1->area() > shape2->area();
+  }
+
+  static bool comparePerimeter(Shape *shape1, Shape *shape2) {
+    return shape1->perimeter() > shape2->perimeter();
+  }
+
+public:
+  // Method to print matching given type
+  void printMatchingType(vector<Shape *> shapes, string name) {
+    for (vector<Shape *>::const_iterator it = shapes.begin();
+         it != shapes.end(); ++it) {
+      if (((*it)->getShapeType()).compare(name) == 0) {
+        cout << (*it)->getShapeType() << endl;
+      }
+    }
+  }
+
+  // Method to print shapes of chosen number of sides
+  void printShapesMatchingNumSides(vector<Shape *> shapes, int numSides) {
+    for (vector<Shape *>::const_iterator it = shapes.begin();
+         it != shapes.end(); ++it) {
+      if (((*it)->getNumSides()) == numSides) {
+        cout << (*it)->getShapeType() << endl;
+      }
+    }
+  }
+
+  // Print the shapes by area descending
+  void printAreaDescending(vector<Shape *> shapes) {
+    vector<Shape *> sortedAreas = shapes;
+    sort(sortedAreas.begin(), sortedAreas.end(), ShapeSorter::compareArea);
```

Comment:
passing the function predicate as a named function is fine, you could also have done this with a lambda, though I prefer your approach for readability.